File: agent1/Backend/app/Services/DashboardService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Models\Alert;
use App\Models\HealthLog;
use App\Models\Machine;
use App\Models\MachineCurrentStatus;
use Exception;
use Illuminate\Support\Facades\Log;

/**
 * Class DashboardService
 *
 * Aggregates dashboard data for company-level and employee-level views.
 * Provides chart data for CPU, RAM, and alert trends.
 *
 * @package App\Services
 */
class DashboardService
{
    /**
     * DashboardService constructor.
     */
    public function __construct()
    {
    }

    /**
     * Get the company dashboard data including summary cards and chart data.
     *
     * Returns an array with total machines, online/offline counts, total alerts,
     * critical alerts, and chart datasets.
     *
     * @param  int    $companyId
     * @return array  ['cards' => [...], 'cpu_chart' => [...], 'ram_chart' => [...], 'alert_chart' => [...]]
     */
    public function getCompanyDashboard(int $companyId): array
    {
        try {
            $machineCounts = Machine::selectRaw('COUNT(*) as total, SUM(CASE WHEN is_online THEN 1 ELSE 0 END) as online_count')
                ->where('company_id', $companyId)
                ->first();

            $alertCounts = Alert::selectRaw("COUNT(*) as total, SUM(CASE WHEN severity = 'critical' AND status IN ('open','acknowledged') THEN 1 ELSE 0 END) as critical_count")
                ->where('company_id', $companyId)
                ->first();

            $cards = [
                'total_machines'   => (int) ($machineCounts->total ?? 0),
                'online_count'     => (int) ($machineCounts->online_count ?? 0),
                'offline_count'    => ((int) ($machineCounts->total ?? 0)) - ((int) ($machineCounts->online_count ?? 0)),
                'total_alerts'     => (int) ($alertCounts->total ?? 0),
                'critical_alerts'  => (int) ($alertCounts->critical_count ?? 0),
            ];

            $chartData = $this->getCombinedChartData($companyId, 24);
            $alertChart = $this->getAlertChartData($companyId, 7);

            return [
                'cards'       => $cards,
                'cpu_chart'   => $chartData['cpu'],
                'ram_chart'   => $chartData['ram'],
                'alert_chart' => $alertChart,
            ];
        } catch (Exception $e) {
            Log::error('DashboardService::getCompanyDashboard - Failed to build dashboard', [
                'company_id' => $companyId,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Get the employee dashboard showing their assigned machine data.
     *
     * @param  int     $userId
     * @return array   ['machine' => Machine|null, 'current_status' => MachineCurrentStatus|null, 'recent_alerts' => Collection]
     */
    public function getEmployeeDashboard(int $userId): array
    {
        try {
            $machine = Machine::with(['currentStatus'])
                ->where('user_id', $userId)
                ->first();

            $recentAlerts = collect();
            if ($machine) {
                $recentAlerts = Alert::where('machine_id', $machine->id)
                    ->orderBy('created_at', 'desc')
                    ->take(10)
                    ->get();
            }

            return [
                'machine'        => $machine,
                'current_status' => $machine?->currentStatus,
                'recent_alerts'  => $recentAlerts,
            ];
        } catch (Exception $e) {
            Log::error('DashboardService::getEmployeeDashboard - Failed to build employee dashboard', [
                'user_id' => $userId,
                'error'   => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Get CPU usage chart data for a company over a specified number of hours.
     *
     * Returns an array of labels and datasets for chart rendering.
     *
     * @param  int  $companyId
     * @param  int  $hours
     * @return array
     */
    public function getCpuChartData(int $companyId, int $hours = 24): array
    {
        return $this->getCombinedChartData($companyId, $hours)['cpu'];
    }

    public function getRamChartData(int $companyId, int $hours = 24): array
    {
        return $this->getCombinedChartData($companyId, $hours)['ram'];
    }

    /**
     * PERFORMANCE OPTIMIZED: Uses SQL-level hourly aggregation instead of loading
     * up to 50,000 raw rows into PHP. Cached for 5 minutes to prevent repeated
     * expensive queries on dashboard refresh.
     *
     * Before: SELECT * FROM health_logs WHERE ... LIMIT 50000 → PHP loop → ~2s
     * After:  SELECT AVG(...) GROUP BY hour, machine_id → ~50ms (cached: ~1ms)
     */
    public function getCombinedChartData(int $companyId, int $hours = 24): array
    {
        try {
            // PERFORMANCE: Cache chart data for 5 minutes — dashboard charts
            // don't need real-time precision, and this query is expensive.
            $cacheKey = "dashboard_chart_{$companyId}_{$hours}";
            $cached = cache()->get($cacheKey);
            if ($cached !== null) {
                return $cached;
            }

            $since = now()->subHours($hours);

            // PERFORMANCE: SQL-level aggregation groups by hour per machine,
            // returning ~200 pre-aggregated points instead of 50,000 raw rows.
            // Uses AVG to smooth the data within each hour bucket.
            $aggregated = HealthLog::selectRaw("
                    machine_id,
                    DATE_FORMAT(collected_at, '%Y-%m-%d %H:00') as hour_bucket,
                    DATE_FORMAT(collected_at, '%H:00') as time_label,
                    ROUND(AVG(cpu_percentage), 1) as avg_cpu,
                    ROUND(AVG(ram_percentage), 1) as avg_ram
                ")
                ->where('company_id', $companyId)
                ->where('collected_at', '>=', $since)
                ->where(function ($q) {
                    $q->whereNotNull('cpu_percentage')
                      ->orWhereNotNull('ram_percentage');
                })
                ->groupBy('machine_id', 'hour_bucket', 'time_label')
                ->orderBy('hour_bucket')
                ->limit(500) // PERFORMANCE: Hard cap to prevent runaway results
                ->get();

            // PERFORMANCE: Single query to get machine names instead of N+1 with()
            $machineIds = $aggregated->pluck('machine_id')->unique()->toArray();
            $machineNames = Machine::select(['id', 'hostname', 'device_name'])
                ->whereIn('id', $machineIds)
                ->get()
                ->keyBy('id');

            $cpuLabelsSet = [];
            $cpuMachineData = [];
            $ramLabelsSet = [];
            $ramMachineData = [];

            foreach ($aggregated as $row) {
                $timeLabel = $row->time_label;
                $m = $machineNames->get($row->machine_id);
                $machineName = $m
                    ? ($m->hostname ?: $m->device_name ?: "Machine {$row->machine_id}")
                    : 'Unknown';

                if ($row->avg_cpu !== null) {
                    $cpuLabelsSet[$timeLabel] = true;
                    $cpuMachineData[$machineName][$timeLabel] = (float) $row->avg_cpu;
                }

                if ($row->avg_ram !== null) {
                    $ramLabelsSet[$timeLabel] = true;
                    $ramMachineData[$machineName][$timeLabel] = (float) $row->avg_ram;
                }
            }

            $cpuLabels = array_keys($cpuLabelsSet);
            $ramLabels = array_keys($ramLabelsSet);

            $buildDatasets = fn($machineData, $labels, $suffix) => array_map(
                fn($name, $data) => [
                    'label' => "{$name} {$suffix}",
                    'data'  => array_map(fn($l) => $data[$l] ?? null, $labels),
                ],
                array_keys($machineData),
                array_values($machineData)
            );

            $result = [
                'cpu' => [
                    'labels'   => $cpuLabels,
                    'datasets' => $cpuMachineData ? $buildDatasets($cpuMachineData, $cpuLabels, 'CPU %') : [['label' => 'No Data', 'data' => []]],
                ],
                'ram' => [
                    'labels'   => $ramLabels,
                    'datasets' => $ramMachineData ? $buildDatasets($ramMachineData, $ramLabels, 'RAM %') : [['label' => 'No Data', 'data' => []]],
                ],
            ];

            // PERFORMANCE: Cache for 5 minutes
            cache()->put($cacheKey, $result, 300);

            return $result;
        } catch (Exception $e) {
            Log::error('DashboardService::getCombinedChartData - Failed', [
                'company_id' => $companyId,
                'hours'      => $hours,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Get alert trend chart data for a company over a specified number of days.
     *
     * Returns daily alert counts grouped by severity.
     *
     * @param  int  $companyId
     * @param  int  $days
     * @return array
     */
    public function getAlertChartData(int $companyId, int $days = 7): array
    {
        try {
            $since = now()->subDays($days)->startOfDay();

            $counts = Alert::selectRaw('DATE(created_at) as date, severity, COUNT(*) as count')
                ->where('company_id', $companyId)
                ->where('created_at', '>=', $since)
                ->groupByRaw('DATE(created_at), severity')
                ->orderBy('date')
                ->get()
                ->keyBy(fn($row) => $row->date . '|' . $row->severity);

            $labels = [];
            $critical = [];
            $warning = [];
            $info = [];

            for ($i = 0; $i <= $days; $i++) {
                $date = now()->subDays($days - $i)->format('Y-m-d');
                $labels[] = now()->subDays($days - $i)->format('D M d');

                $critical[] = (int) ($counts->get("{$date}|critical")?->count ?? 0);
                $warning[]  = (int) ($counts->get("{$date}|warning")?->count ?? 0);
                $info[]     = (int) ($counts->get("{$date}|info")?->count ?? 0);
            }

            return [
                'labels' => $labels,
                'datasets' => [
                    ['label' => 'Critical', 'data' => $critical],
                    ['label' => 'Warning',  'data' => $warning],
                    ['label' => 'Info',     'data' => $info],
                ],
            ];
        } catch (Exception $e) {
            Log::error('DashboardService::getAlertChartData - Failed to get alert chart data', [
                'company_id' => $companyId,
                'days'       => $days,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }
}

File: DeskGuardBackend/Services/DashboardService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Logging;
using DeskGuardBackend.Data;
using DeskGuardBackend.Entities;
using DeskGuardBackend.Services.Interfaces;

namespace DeskGuardBackend.Services
{
    public class DashboardService : IDashboardService
    {
        private readonly DeskGuardDbContext _dbContext;
        private readonly IMemoryCache _cache;
        private readonly ILogger<DashboardService> _logger;

        public DashboardService(
            DeskGuardDbContext dbContext,
            IMemoryCache cache,
            ILogger<DashboardService> logger)
        {
            _dbContext = dbContext;
            _cache = cache;
            _logger = logger;
        }

        public async Task<object> GetCompanyDashboardAsync(long companyId)
        {
            try
            {
                // Machine counts
                var totalMachines = await _dbContext.Machines.CountAsync(m => m.CompanyId == companyId);
                var onlineMachines = await _dbContext.Machines.CountAsync(m => m.CompanyId == companyId && m.IsOnline);

                // Alert counts (total & critical open/acknowledged)
                var totalAlerts = await _dbContext.Alerts.CountAsync(a => a.CompanyId == companyId);
                var criticalAlerts = await _dbContext.Alerts
                    .CountAsync(a => a.CompanyId == companyId && a.Severity == "critical" && (a.Status == "open" || a.Status == "acknowledged"));

                var cards = new
                {
                    total_machines = totalMachines,
                    online_count = onlineMachines,
                    offline_count = totalMachines - onlineMachines,
                    total_alerts = totalAlerts,
                    critical_alerts = criticalAlerts
                };

                var chartData = await GetCombinedChartDataInternalAsync(companyId, 24);
                var alertChart = await GetAlertChartDataAsync(companyId, 7);

                return new
                {
                    cards = cards,
                    cpu_chart = chartData.Cpu,
                    ram_chart = chartData.Ram,
                    alert_chart = alertChart
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "DashboardService::GetCompanyDashboardAsync failed for company: {CompanyId}", companyId);
                throw;
            }
        }

        public async Task<object> GetEmployeeDashboardAsync(long userId)
        {
            try
            {
                var machine = await _dbContext.Machines
                    .Include(m => m.CurrentStatus)
                    .FirstOrDefaultAsync(m => m.UserId == userId);

                var recentAlerts = new List<Alert>();
                if (machine != null)
                {
                    recentAlerts = await _dbContext.Alerts
                        .Where(a => a.MachineId == machine.Id)
                        .OrderByDescending(a => a.CreatedAt)
                        .Take(10)
                        .ToListAsync();
                }

                return new
                {
                    machine = machine,
                    current_status = machine?.CurrentStatus,
                    recent_alerts = recentAlerts
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "DashboardService::GetEmployeeDashboardAsync failed for user: {UserId}", userId);
                throw;
            }
        }

        public async Task<object> GetCpuChartDataAsync(long companyId, int hours = 24)
        {
            var data = await GetCombinedChartDataInternalAsync(companyId, hours);
            return data.Cpu;
        }

        public async Task<object> GetRamChartDataAsync(long companyId, int hours = 24)
        {
            var data = await GetCombinedChartDataInternalAsync(companyId, hours);
            return data.Ram;
        }

        public async Task<object> GetAlertChartDataAsync(long companyId, int days = 7)
        {
            try
            {
                var since = DateTime.UtcNow.AddDays(-days).Date;

                var alertCounts = await _dbContext.Alerts
                    .Where(a => a.CompanyId == companyId && a.CreatedAt >= since)
                    .GroupBy(a => new { Date = a.CreatedAt.Date, Severity = a.Severity })
                    .Select(g => new
                    {
                        g.Key.Date,
                        g.Key.Severity,
                        Count = g.Count()
                    })
                    .ToListAsync();

                var labels = new List<string>();
                var critical = new List<int>();
                var warning = new List<int>();
                var info = new List<int>();

                for (int i = 0; i <= days; i++)
                {
                    var date = DateTime.UtcNow.AddDays(-days + i).Date;
                    labels.Add(date.ToString("ddd MMM dd"));

                    critical.Add(alertCounts.FirstOrDefault(c => c.Date == date && c.Severity == "critical")?.Count ?? 0);
                    warning.Add(alertCounts.FirstOrDefault(c => c.Date == date && c.Severity == "warning")?.Count ?? 0);
                    info.Add(alertCounts.FirstOrDefault(c => c.Date == date && c.Severity == "info")?.Count ?? 0);
                }

                return new
                {
                    labels = labels,
                    datasets = new[]
                    {
                        new { label = "Critical", data = critical },
                        new { label = "Warning", data = warning },
                        new { label = "Info", data = info }
                    }
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "DashboardService::GetAlertChartDataAsync failed for company: {CompanyId}", companyId);
                throw;
            }
        }

        private class CombinedChartData
        {
            public object Cpu { get; set; } = null!;
            public object Ram { get; set; } = null!;
        }

        private async Task<CombinedChartData> GetCombinedChartDataInternalAsync(long companyId, int hours)
        {
            var cacheKey = $"dashboard_chart_{companyId}_{hours}";

            if (_cache.TryGetValue<CombinedChartData>(cacheKey, out var cachedData) && cachedData != null)
            {
                return cachedData;
            }

            try
            {
                var since = DateTime.UtcNow.AddHours(-hours);

                // Group by Machine, Date and Hour to aggregate hourly
                var aggregatedData = await _dbContext.HealthLogs
                    .Where(h => h.CompanyId == companyId && h.CollectedAt >= since && (h.CpuPercentage != null || h.RamPercentage != null))
                    .GroupBy(h => new
                    {
                        h.MachineId,
                        Date = h.CollectedAt!.Value.Date,
                        Hour = h.CollectedAt!.Value.Hour
                    })
                    .Select(g => new
                    {
                        g.Key.MachineId,
                        g.Key.Date,
                        g.Key.Hour,
                        AvgCpu = g.Average(x => x.CpuPercentage),
                        AvgRam = g.Average(x => x.RamPercentage)
                    })
                    .OrderBy(x => x.Date)
                    .ThenBy(x => x.Hour)
                    .Take(500)
                    .ToListAsync();

                // Format hour buckets in memory
                var aggregated = aggregatedData.Select(x => new
                {
                    x.MachineId,
                    HourBucket = new DateTime(x.Date.Year, x.Date.Month, x.Date.Day, x.Hour, 0, 0, DateTimeKind.Utc),
                    x.AvgCpu,
                    x.AvgRam
                }).ToList();

                var machineIds = aggregated.Select(x => x.MachineId).Distinct().ToList();
                var machines = await _dbContext.Machines
                    .Where(m => machineIds.Contains(m.Id))
                    .ToDictionaryAsync(m => m.Id, m => m.Hostname ?? m.DeviceName ?? $"Machine {m.Id}");

                var cpuLabelsSet = new SortedSet<string>();
                var ramLabelsSet = new SortedSet<string>();

                var cpuMachineData = new Dictionary<string, Dictionary<string, float>>();
                var ramMachineData = new Dictionary<string, Dictionary<string, float>>();

                foreach (var row in aggregated)
                {
                    var timeLabel = row.HourBucket.ToString("HH:00");
                    machines.TryGetValue(row.MachineId, out var machineName);
                    machineName ??= "Unknown";

                    if (row.AvgCpu.HasValue)
                    {
                        cpuLabelsSet.Add(timeLabel);
                        if (!cpuMachineData.ContainsKey(machineName)) cpuMachineData[machineName] = new Dictionary<string, float>();
                        cpuMachineData[machineName][timeLabel] = (float)Math.Round(row.AvgCpu.Value, 1);
                    }

                    if (row.AvgRam.HasValue)
                    {
                        ramLabelsSet.Add(timeLabel);
                        if (!ramMachineData.ContainsKey(machineName)) ramMachineData[machineName] = new Dictionary<string, float>();
                        ramMachineData[machineName][timeLabel] = (float)Math.Round(row.AvgRam.Value, 1);
                    }
                }

                var cpuLabels = cpuLabelsSet.ToList();
                var ramLabels = ramLabelsSet.ToList();

                var buildDatasets = new Func<Dictionary<string, Dictionary<string, float>>, List<string>, string, object[]>((machineData, labels, suffix) =>
                {
                    return machineData.Select(kvp => new
                    {
                        label = $"{kvp.Key} {suffix}",
                        data = labels.Select(l => kvp.Value.TryGetValue(l, out var val) ? (float?)val : null).ToArray()
                    }).ToArray();
                });

                var result = new CombinedChartData
                {
                    Cpu = new
                    {
                        labels = cpuLabels,
                        datasets = cpuMachineData.Count > 0 ? buildDatasets(cpuMachineData, cpuLabels, "CPU %") : new object[] { new { label = "No Data", data = Array.Empty<float?>() } }
                    },
                    Ram = new
                    {
                        labels = ramLabels,
                        datasets = ramMachineData.Count > 0 ? buildDatasets(ramMachineData, ramLabels, "RAM %") : new object[] { new { label = "No Data", data = Array.Empty<float?>() } }
                    }
                };

                // Cache for 5 minutes
                _cache.Set(cacheKey, result, TimeSpan.FromMinutes(5));

                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "DashboardService::GetCombinedChartDataInternalAsync failed for company: {CompanyId}", companyId);
                throw;
            }
        }
    }
}
